helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -544,6 +544,19 @@ public synchronized void setLiveInstances(List<LiveInstance> liveInstances) {
     return _instanceMessagesCache.getMessages(instanceName);
   }
 
+  public Map<String, Map<String, Message>> getStaleMessages() {
```

Comment:
So this one will return the map for modifying. I still don't like this implicit map/list ref for modifying purposes.
As I checked the current usage, I think you can replace it with 2 explicit modifying methods,
1. addStaleMessage(instance, message)
2. removeStaleMessage(instance, message) -- or maybe always clean up when the cache is refreshed, so you don't even need this method?

As for supporting test, we can make it protected so only visible to test. Or keeping this method but always return shallow copy of the map only for the tests.